---
title: "RNAseq_Analysis"
author: "Tzu L. Phang"
date: "April 29, 2015"
output: 
  pdf_document:
    latex_engine: lualatex
---

# RNAseq Analysis

## Background

This tutorial outline an RNAseq analysis routine conducted in my R + Biocondcutor data analysis course (BIOS6660) using one of my collaborator's data set, Dr. Eric Schmidt.  Since RNAseq dataset is typically very large, we extracted only Chromosome 19 to ease perform the analysis on local computer (mostly laptops)

## Experimental Design

Read in the descriptive data file and run summary statistics

```{r, cache=TRUE, results='markup', message=FALSE, warning=FALSE}
## Read in descriptive data file
targets = read.table('./data/targets.txt', header = T)
## printing file table
knitr::kable(targets)
```

It seems like we have a total of 6 FASTQ files and 2 experimental grouping (CLP vs. Sham) in mouse sample


## Mapping FASTQ to genome

We will be using QuasR, which is an extremely versatile NGS mapping and postprocessing pipeline for RNA-seq It uses Rbowtie for upgapped alignments adn SpliceMap for spliced alignments

```{r, cache=TRUE, results='markup', message=FALSE, warning=FALSE}
library(QuasR)
targets = read.table("./data/targets.txt", header = T)
write.table(targets[,1:2], 'data/QuasR_samples.txt', row.names=F, quote=F, sep='\t')
sampleFile = "./data/QuasR_samples.txt"
genomeFile = "./data/Mouse.chromosome.19.fa"
results = "./result"
cl = makeCluster(10)


## Single command to index reference, align all samples and generate BAM files
proj <- qAlign(sampleFile, genome=genomeFile, maxHits=1, splicedAlignment=T, alignmentsDir=results, 
               clObj=cl, cacheDir=results)
# Note: splicedAlignment should be set to TRUE when the reads are >=50nt long  
alignstats <- alignmentStats(proj) # Alignment summary report
alignstats = as.data.frame(alignstats)
align.mapped.percent = alignstats$

knitr::kable(alignstats)
```
 
 

